### ISSUE TITLE
Guard with a boolean doesn't work?

### ISSUE THREAD

**Issue by batiste** [facebook/flow#5755]
https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVUCuA7AxgFwEs5sx8BTAZ3wAoALcmeALjGoCdDsBzMAHzAAjOHBjkAhtgCUYAN6owYXCWphumcewAmYALxgGTOArCEoNdZq0z5ixe3L5M7UgANuDx114ASWYfgAX1cTQNRAoA

```javascript
/* @flow */

function test(hello: string | boolean) {
  const guard = hello
  if(guard) {
    return `greeting ${hello}`
  }
}
```

**Comment by apsavin** [comment]
It's not the answer, but you can try to set `experimental.const_params=true` in your flow config.

**Comment by TrySound** [comment]
Or you may use
```
// @flow strict
```
for the same effect

**Comment by batiste** [comment]
Is really the answer? Flow cannot handle such simple variable renaming? I am flabbergasted!

**Comment by TrySound** [comment]
@batiste I think it's for the best. Code will be cleaner.

**Comment by batiste** [comment]
Cleaner? I mean this is a simplified example here is a little more what would happen in reality
What is the code smell you would like to avoid? Variable renaming for clarity seems like essential and basic no?

```javascript
/* @flow */

function test(hello?: string) {
  const guard = hello || "something"
  if(guard) {
    return `greeting ${hello}`
  }
}
```

**Comment by TrySound** [comment]
@batiste In this example flow prevents you from passing `undefined`. You don't only rename variable, but also changed its type.

And variable renaming without changing value is something just weird and unnecessary.

**Comment by batiste** [comment]
@TrySound yes I do change the type of guard and ensure this is a string which is different from the example before. But why does flow still fail in this case? Even with 2 precautionary mechanisms? Wouldn't you say that flow should be smart enough to see that guard HAS to be a string at this point?
It just seems to me that flow gives up after a variable rename... How weird...

**Comment by batiste** [comment]
@TrySound ha yeah I just realised my mistake, this works fine for example:

```javascript
/* @flow */

function test(hello?: string) {
  const guard = hello || "something"
  if(guard) {
    return `greeting ${guard}`
  }
}
```

As well as this

```javascript
/* @flow */

function test(hello?: string) {
  const guard = hello
  if(guard) {
    return `greeting ${guard}`
  }
}
```

I can understand why the flow compiler could somehow confused... Although it still weird it is unable to realizer guard is strictly equal to hello

**Comment by batiste** [comment]
@TrySound yeah sure. The example I gave at the beginning was a highly distilled code from a bigger code function that seems to make sense at the time. I will ask my colleague to maybe produce an example that is more representative @lukaszwado maybe you have an idea?

**Comment by lukaszwado** [comment]
Sorry for late response. 
I think the original problem was something like: 

```
function test(hello?: string) {
  const isHelloValid = typeof hello !== 'undefined' && hello.length > 0

  if(isHelloValid) {
    return `greeting ${hello}`
  }
}
```

In order to satisfy flow I would need to do my check twice i.e. 

```
function test(hello?: string) {
  const isHelloValid = typeof hello !== 'undefined' && hello.length > 0

  if(isHelloValid &&  typeof hello !== 'undefined') {
    return `greeting ${hello}`
  }
}
```
